### PR TITLE
UserID module: better initialization logic

### PR DESCRIPTION
**Pull request by dgirardi**
## Type of change

- [x] Bugfix
- [x] Does this change affect user-facing APIs or examples documented on http://prebid.org?

## Description of change

﻿This updates the userID module to avoid premature initialization (https://github.com/prebid/Prebid.js/issues/5447):

 - `pbjs.getUserIds` and `pbjs.getUserIdsAsEids` no longer force initialization of the ID system
 - ID submodules are now initialized after after GDPR consent has been resolved *and* `userSync` configuration has been set
 - added `pbjs.getUserIdsAsync()`, which returns a promise to user IDs, completing only once all ID submodules have completed init
 - `pbjs.refreshUserIds` now forces an immediate refresh regardless of `userSync` config and returns a promise that completes when all ID submodules have filled in their data. `pbjs.refreshUserIds({refresh: false})` will just return the promise without refreshing.

Note that until https://github.com/prebid/Prebid.js/pull/8185 gets merged, GDPR data cannot be resolved without an auction, so this will prevent any ID from being resolved if the consentManagement module is installed and no auction is started. No additional action should be required to fix this besides the merge.

## Other information

Fixes https://github.com/prebid/Prebid.js/issues/5447
Documentation PR: https://github.com/prebid/prebid.github.io/pull/3698

### CODE REVIEW

**Comment by patmmccann** [comment]
@dgirardi please note the merge conflict

**Comment by EskelCz** [comment]
Thanks for this, looks promising.
One suggestion for readability though, `pbjs.refreshUserIds({refresh: false})` is just what we want, but it's not immediately clear what it does. Can it separated into a new method, something like `pbjs.getUserIdsAsync()`?

**Comment by dgirardi** [comment]
What about adding `getUserIdsAsync` but removing callbacks from the other methods' signatures? Since all of these are part of the public API I'd like to keep changes to a minimum.

**Comment by EskelCz** [comment]
Sounds good, that would leave the current implementations that are working as they are... and whoever is having issues would switch to the async version.

**Review by smenzer — CHANGES_REQUESTED** [review]
Overall this looks good. I'm not an engineer so I can't say that every bit of code is written in the best way, but the logic makes sense and I like that it doesn't break existing implementations.

I would like to see the examples page updated to show how to properly use the `pbjs.getUserIds()` method with the new callback function, as well as any related docs (I know the docs PR is still pending). Once all of that is done, I'm happy to approve. 

We will need another engineer to review and approve since this is core.

File: test/spec/modules/userId_spec.js
Comment (smenzer) [inline test/spec/modules/userId_spec.js:155]: I agree completely. a lot of this was from the first iteration of the user id module, so a lot of copy/paste from newer modules and signatures weren't necessarily updated as we made changes.

I think we need to have a few tests that do test *all* id systems for basic processing, but the majority of these tests could probably just be mock id systems to test core functionality

**Review by smenzer — APPROVED** [review]

**Review by SKOCHERI — APPROVED** [review]

**Comment by smenzer** [comment]
We've got 2 approvals on this, so I'm going to go ahead and merge it in. Thanks for all the work on this one everyone, especially @dgirard and @EskelCz!

**Comment by EskelCz** [comment]
@smenzer @dgirardi Awesome, thanks a lot for quick resolution. Looking forward to testing it out.

**Pull request merged**
